### DISCUSSION TITLE
Add the Ability for "custom" release dates

### DISCUSSION THREAD

**Discussion by Kiki79250CoC** [community/community#6646] [Other Feature Feedback, Questions, & Ideas] [answered]
Hi,

I write this message to request if it is possible to add "Custom" release date.

> **_Context :_**
> A few months ago, I decided to publish the source code on GitHub for an existing project. This project already had quite a few versions that were released before I created the repository.
>
> When I uploaded these versions (which was released between December 2018 and June 2021), the release date that is shown is the upload date, not the original release date.
>
> Before this UI update, I placed the actual release date on the title in their descriptions. But now since the update I'm forced to do it in the main title.

Examples of two releases I released before Repo creation and I uploaded in it when I created it :
![image](https://user-images.githubusercontent.com/54137141/138541742-4e3be5e1-ed92-4d40-b763-c5eb7edea9a2.png)
![image](https://user-images.githubusercontent.com/54137141/138541871-8e81172e-7b49-46cc-959e-ecf6d21c2ddc.png)

This problem disappears for the newer versions I released after the repo creation, but for older releases, it's not very clear...

So if i would set (for example of this old release) "Released on December 28, 2018" instead of "3 months ago", this will avoid me to add the release date in the title itself.

-----

For the UI, I thinked to add a little calendar box labelled "Release Date", settled by default on the current date, like on this little photo montage : 
![image](https://user-images.githubusercontent.com/54137141/138542833-b98777d7-dcdb-42c8-9bec-f1c7c25a8553.png)

And it will appear like this on the "Releases" Page :
![image](https://user-images.githubusercontent.com/54137141/138542796-18b177a6-6954-429c-bfeb-041b4626bada.png)

I think it will be better like this.

Regards,
Kylian

**Comment by MylesBorins** [comment, marked as answer]
Hey @Kiki79250CoC thanks for starting this discussion. Dates for releases is... tricky.

You have at least 2 dates associated with the release, and potentially as many as 3!

You have:

* The date of the commit
* The date of the tag
* The date of the GitHub Release

Here's where things get tricky. By default a tag is not annotated ([docs](https://git-scm.com/docs/git-tag#_description))... this means the tag has no specific meta data associated with it (such as author or date) and is just a loose pointer to a commit. You need to run tag with `-a`, `-s`, `-f,` or `-m` to make an annotated tag that has all of this additional information.

So here is where it gets fun! Let's say you made a release candidate, and that RC is the exact same commit as your release, but you use non annotated tags. On our end we actually don't know the difference between the two tags aside from the fact that they have different names, and depending on the way you are doing alphabetical sort 1.2.3-rc.1 could sort above 1.2.3... not the outcome you want!

So we use the release date of the GitHub release to be the canonical date of the "GitHub release", but it is not meant to be a substitution for the tag or commit. In fact we use the tag / commit date to sort releases rather than the GitHub Release date for the exact reason you mentioned above!

That said, if someone made a release candidate and promoted it to be a full release using non annotated tags we could end up with a date that is before the actual release.

On to your suggested solution. We could allow for a customized release date, but considering we already have 3 different ways of making the date, and that you can customize [the date of the tags itself](https://git-scm.com/docs/git-tag#_date_formats) I don't think it would be a good idea for us to introduce a 4th possible date.

We do show the GitHub Release date as the primary date in the feed and on the release itself, which is something that we could explore changing... but for the 99% case the GitHub Release date is likely to be the expected date of the release, and if you think of it like a blog post, the publish day of the post is the "day" of the GitHub release.

Anyways... apologies for the many many words to say "it's complicated"... but unfortunately like most times when dealing with computers, dates are hard 🙃
